Add line graph to draft applications report

PaulMorris opened the pull request:
These changes are made on top of those in PR #812 and won't be ready to review until #812 is merged.  Also, this is still WIP, so here's a WIP screenshot:

![screenshot-2018-5-11 draft report](https://user-images.githubusercontent.com/1091693/39947793-351c82b0-5541-11e8-8160-bbce1a3db99d.png)

Issue #783: Report for number of applications left in draft status

cecilia-donnelly commented:
@PaulMorris, now that #812 is merged can this be rebased?  Excited about the possibilities shown by the screenshot!

cecilia-donnelly commented on psm-app/frontend/src/main/js/admin/report.js:
```diff
@@ -1,3 +1,5 @@
+"use strict";
+
 function reportTableToJson (table) {
   data = [];
```
I know this is WIP, but just so you know Firefox yelled at me for assigning a value to "undefined variable data."  I fixed it by writing `var data = [];` instead.

PaulMorris replied:
Good catch!  Fixed.

PaulMorris commented:
@cecilia-donnelly, @frankduncan   Now rebased on current master and ready for review.  I think we could go ahead and merge it as-is and improve it in follow-up PRs (the other option being to improve it in this PR before merging).

frankduncan commented on psm-app/cms-web/WebContent/WEB-INF/pages/admin/reports/draft_applications_page.jsp:
```diff
@@ -69,7 +72,5 @@
-  <script>
-    var reportInformation = reportTableToJson($(".reportTable"));
-  </script>
+  <script src="${ctx}/js/admin/draftsReport.js"></script>
```
Do we prefer this to doing an onload with the script at the top?

PaulMorris replied:
Now that you mention it, I think having an onload handler and keeping the scripts at the top is a better way to go.  I'll do that at least, and take a look at abstracting the line graph code into a function that can be used on multiple pages.

cecilia-donnelly commented on psm-app/frontend/src/main/js/admin/draftsReport.js:
```diff
@@ -0,0 +1,103 @@
+});
+
+var firstMonth = d3.min(months);
+var lastMonth = d3.max(months);
```
So right now the graph will show months over all time?  What about if we have multiple years worth of Januarys -- are the counts of all Januarys combined into one number?

PaulMorris replied:
The month values are specific to a given year, namely a string, e.g. '2018-05-01'.  I'll add a comment to make this clear.

Currently the graph shows all the months that are in the html table on the page.  Making the range of months user-modifiable is still on the to do list.

cecilia-donnelly commented on psm-app/frontend/src/main/js/admin/draftsReport.js:
```diff
@@ -0,0 +1,103 @@
+
+var reportJson = reportTableToJson($(".reportTable"));
+
+var monthCounts = reportJson.reduce(function(result, item) {
```
I'd appreciate a comment here explaining the expected format of `result` and `item`, and ideally more descriptive variable names, since these are quite generic.

PaulMorris replied:
Okay, I'll add a comment and change the names.  (The genericness is in part that functions passed to reduce in JS all have this signature, where the first argument is the result you're building up, and the second is that item in the array.  But more specific names would be easier to read.)

cecilia-donnelly commented on psm-app/frontend/src/main/js/admin/draftsReport.js:
```diff
@@ -0,0 +1,103 @@
+// Rendering line graph
+
+var WIDTH = 960;
+var HEIGHT = 460;
```
This makes the graph not at all responsive -- can we use percentages here instead?  It's a fine size on a large monitor, but it might be unusably big on a smaller screen.

cecilia-donnelly replied:
(Definitely lower priority if it's hard, but nice to have if it's pretty easy to change.)

PaulMorris replied:
Good call.  The pages that contain the graph are currently also non-responsive, with a set width, so I just went with that.  But I can try percentages on width.  Height might be less straightforward...

cecilia-donnelly commented on psm-app/frontend/src/main/js/admin/draftsReport.js:
```diff
@@ -0,0 +1,103 @@
+
+var yScale = d3
+  .scaleLinear()
+  .domain([0, maxDrafts + maxDrafts * 0.1])
```
Nice.

cecilia-donnelly commented on psm-app/frontend/src/main/js/admin/report.js:
```diff
@@ -10,3 +12,36 @@ function reportTableToJson (table) {
   return data;
 }
+
+// Date arithmetic utilities
```
Please add doc strings for these explaining what they do, the expected values of their args, etc.

cecilia-donnelly commented:
I agree with @frankduncan's comment " it may behoove us to break out the code into a form that can be called with one line specifying the various fields the graph can look at."  That would be super useful.  It'll need to happen for the second line graph you make anyway, though, so up to you whether you think it belongs in this PR or the next one.

PaulMorris commented:
@cecilia-donnelly @frankduncan This revision addresses your review comments, and pulls the code out into a couple of reusable functions for line graphs.  

I held off on responsiveness for now, since the width and height values are used to calculate other dimensions of the graph (when combined with margins).  So that change was not going to be straightforward.

Using the window 'load' handler added some indirection to get the script into the html head element (via the jsp for the head), but there's some value to having all the scripts added in the same way in one place.  I'm not thrilled with this, but I think it's better than having script tags at the bottom of the page.  (If the html head jsp just had the contents of the head tag, but not the opening and closing head tags themselves, you could add scripts on the fly to the head of a given page and avoid the indirection, but that might or might not be better and it's getting way out of scope for this PR!)

PaulMorris commented:
New screenshot with the seed.sql data:

![screenshot-2018-5-17 draft applications](https://user-images.githubusercontent.com/1091693/40200921-9502a30e-59eb-11e8-8e67-3d0113a89974.png)

frankduncan commented:
@PaulMorris Thanks!  This looks good.  I have one last question, mainly for discussion, and then I think it's good to merge!

PaulMorris commented on psm-app/cms-web/WebContent/templates/includes/html_head.template.html:
```diff
@@ -62,4 +62,8 @@
+
+  {{#if draftsReport}}
+    <script src="{{ctx}}/js/admin/draftsReport.js"></script>
+  {{/if}}
```
Yeah, I wasn't thrilled about the approach here.  I like your idea much better.  I'll see if I can make that work in handlebars.

PaulMorris commented:
@frankduncan It was taking too long trying to define a list of urls in JSP syntax and then get handlebars to iterate over them (sigh...) so I just kept it simple -- passing a single string for a given script to add to the page.  This is much better than the previous version at least.

frankduncan commented:
@PaulMorris I took a shot at doing the multiple files, and came up with the changes in #2c6bdfa

If that looks tenable, we can keep it in

PaulMorris commented:
@frankduncan Looks good, thanks!

PaulMorris commented:
@frankduncan FYI, In the interest of time I went ahead and modified your commit to remove the handlebars `if`.  (Hope you don't mind.)  And it looks like my comments on your commit are no longer showing up above (since I amended the commit and force pushed).  Basically, we don't need the `if` because `each` already gives us the results we want without it.  (If `pageScripts` is not defined, then nothing happens.)

So this is ready to merge.  @cecilia-donnelly do you want to take another look or shall I merge away?

PaulMorris commented:
@jasonaowen This now passes JS lint checks.  Also, once this lands master should be free of linting issues.

cecilia-donnelly commented:
I don't need to be a blocker, @PaulMorris -- if you and @frankduncan are happy then go ahead and merge.

jasonaowen commented:
Thank you, @PaulMorris !

frankduncan commented:
> @frankduncan FYI, In the interest of time I went ahead and modified your commit to remove the handlebars if. (Hope you don't mind.) And it looks like my comments on your commit are no longer showing up above (since I amended the commit and force pushed). Basically, we don't need the if because each already gives us the results we want without it. If pageScripts is not defined, then nothing happens. (I confirmed this by building the PSM and checking the draft reports page. Everything worked fine.)

Oh, definitely don't mind.  Thanks a ton for rocking it!

And in general, yeah, this looks great!  Merge away.
